mlx/fast.h

```diff
@@ -61,4 +62,52 @@ array affine_dequantize(
     int bits = 4,
     StreamOrDevice s = {});
 
+typedef std::variant<int, bool, Dtype> TemplateArg;
+
+class MetalKernel {
+ public:
+  MetalKernel(
+      const std::string& name,
+      const std::string& source,
+      std::unordered_map<std::string, std::vector<int>> output_shapes,
+      std::unordered_map<std::string, Dtype> output_dtypes,
+      std::tuple<int, int, int> grid,
+      std::tuple<int, int, int> threadgroup,
+      bool ensure_row_contiguous,
+      bool verbose)
+      : name_(name),
+        source_(source),
+        output_shapes_(output_shapes),
+        output_dtypes_(output_dtypes),
+        grid_(grid),
+        threadgroup_(threadgroup),
+        ensure_row_contiguous_(ensure_row_contiguous),
+        verbose_(verbose) {
+    validate_output_shapes();
+  }
+
+  std::unordered_map<std::string, array> run(
+      std::unordered_map<std::string, array>& inputs,
+      StreamOrDevice s = {});
```

Comment:
Ideally we can match the C++ and Python APIs. So depending on how this changes it could either be an overload of operator() or you return a lambda from the `metal_kernel` function.